OIDC Auth for Funnel, k8s state support

lstoll opened the pull request:
This adds OIDC auth for funnel usage. By default all funnel requests will require this, unless explicitly overridden.

Because this makes the config more complex, move to just using a file for it.

Add support for persisting state in a k8s secret too.

Add CI build, with docker image.

## TODO
- [ ] Update README
- [ ] Resolve local tailscale daemon usage

lstoll committed:
Move config to file

Going to get more complex with k8s config + OIDC, make it a file.

lstoll committed:
Kubernetes state backend

Want to run it in k8s, add a backend that uses a secret to store the data.
While tailscale has one in their codebase, it's not suitable for our usage so
we just add a simple one.

lstoll committed:
use the hujson power

lstoll committed:
Break out funnel listener

Want a dedicated listener, so we can selectively apply auth etc. So just add a
new listener/server for that.

Redirect to https for internal requests on funnel fqdn's, to keep URLs
consistent.

Rework startup/shutdown a little to be more graceful.

lstoll committed:
OIDC auth for funnel

Adds OIDC auth to funnel endpoints, setting user header consistently.

Optionally paths can be specified to be authless.

lstoll committed:
an example config

lstoll committed:
CI build + docker publish

lstoll committed:
Disable local tailscale usage

When running in a container, tailscale doesn't already exist so we can't talk
to it.

Just use a local listener, need to think about this more.

lstoll commented:
Need to think about the local tailscale daemon usage - it's there to figure out where to bind the discovery and metrics endpoints, but this doesn't work when it's running in a container.

Might make a config option (ugh, checkboxes) for metrics listen - either set it to a fixed ip/port, or fallback to daemon discovery. But will think on it more

lstoll committed:
correct error message

sr commented on k8s_state.go:
```diff
@@ -0,0 +1,117 @@
+package main
```
may be able to use https://pkg.go.dev/tailscale.com/kube:

https://github.com/tailscale/tailscale/blob/main/cmd/containerboot/kube.go#L21-L23

I don't mind client-go tho, it's fine

sr commented on main.go:
```diff
@@ -205,98 +163,210 @@ func tsproxy(ctx context.Context) error {
 
-	for i, upstream := range upstreams {
-		// https://go.dev/doc/faq#closures_and_goroutines
-		i := i
```
hehe nice

sr commented on config.go:
```diff
@@ -0,0 +1,152 @@
+	// OIDCClientID sets the OIDC client ID
+	OIDCClientID string `json:"oidcClientID"`
+	// OIDCClientSecret sets the OIDC client secret
+	OIDCClientSecret string `json:"oidcClientSecret"`
```
be cool to support a file too, play nice with systemd-creds. For flags I usually do this:

```
	if _, path, ok := strings.Cut(*cliSecret, "file://"); ok {
		b, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		s := strings.TrimSpace(string(b))
		cliSecret = &s
	}
```

but doesn't really work w/ a config file. Maybe oidcClientSecretFile?

sr replied:
also wondering if this should be on the root config struct, or do you plan on using a different issuer for some upstream(s)?

sr commented on config.go:
```diff
@@ -0,0 +1,152 @@
+	StateDir string `json:"stateDir"`
+	// Kubernetes configures the proxy to run in a kubernetes cluster. In this
+	// case the StateDir is ignored, and state managed in a secret.
+	Kubernetes kubernetesConfig `json:"kubernetes"`
```
*kubernetesConfig? then can drop the enabled field.

sr commented on main.go:
```diff
@@ -121,64 +79,64 @@ func main() {
+				}
+				listeners = append(listeners, ln)
+			}*/
+		ln, err := net.Listen("tcp", "0.0.0.0:"+p)
```
I use `http://nuc10i5/sd` in my prom config, I guess I could make tsproxy.service listen on the node's tailscale IP using some ansible: `{{ ansible_facts.tailscale0.ipv4.address }`

alternatively could you check for the presence of a local socket? will think about it too.

sr commented on main.go:
```diff
@@ -330,15 +402,35 @@ func newReverseProxy(logger *slog.Logger, lc tailscaleLocalClient, url *url.URL)
+		// Proxy requests from non-funnel tagged nodes as is.
+		//
+		// TODO(lstoll) figure out why - and if end-user nodes would be tagged?
+		if whois.Node.IsTagged() && !isFunnel {
```
yeah I guess a config w/ a list of tags, or maybe node id to consider as end-users??

lstoll committed:
Merge remote-tracking branch 'origin/main' into lstoll-k8s-oidc

lstoll committed:
fix up some merge issues

lstoll committed:
tailscale v1.84.0

lstoll committed:
gofumpt

lstoll committed:
lstoll/oidc @ d4b0188

lstoll committed:
Use updated OIDC middleware

lstoll committed:
bump lint version in CI

lstoll committed:
Make caching work

For it to work, the checkout needs to happen first

lstoll committed:
fix golangci config so it passes with config verify

lstoll committed:
Update lstoll/oidc with invalid token fix

lstoll committed:
go mod tidy

lstoll committed:
Restore public paths not requiring OIDC token

lstoll committed:
Update for oauth2ext rename

lstoll committed:
Update dependencies

lstoll committed:
Dependency updates

* Switch to lds.li/oauth2ext import path, get latest
* go get -u ./...
* go get gvisor.dev/gvisor@9414b50a5633 (from tailscale go.mod)

lstoll committed:
Add support for client registration + group enforcement

If configured, dynamically register and renew a client.

Add support for filtering groups to allow access.

lstoll committed:
Update oauth2ext
